WMF Framework/Event Platform/StorageManager.swift

```diff
@@ -0,0 +1,185 @@
+
+        do {
+            try psc.addPersistentStore(ofType: NSSQLiteStoreType, configurationName: nil, at: storageURL, options: options)
+        } catch {
+            do {
+                try FileManager.default.removeItem(at: storageURL)
+            } catch {
+
+            }
+            do {
+                try psc.addPersistentStore(ofType: NSSQLiteStoreType, configurationName: nil, at: storageURL, options: options)
+            } catch {
+                DDLogError("EPC: Event Platform StorageManager: adding persistent store to coordinator: \(error)")
+                return nil
+            }
+        }
+
+        let managedObjectContext = NSManagedObjectContext(concurrencyType: .privateQueueConcurrencyType)
+        managedObjectContext.persistentStoreCoordinator = psc
+        self.managedObjectContext = managedObjectContext
+    }
+
+    private func pruneStaleEvents() {
+        perform { moc in
+            let pruneFetch = NSFetchRequest<NSFetchRequestResult>(entityName: "WMFEPEventRecord")
+            pruneFetch.returnsObjectsAsFaults = false
+
+            let pruneDate = Date().addingTimeInterval(-(self.pruningAge)) as NSDate
+            pruneFetch.predicate = NSPredicate(format: "(recorded < %@) OR (posted != nil) OR (failed == TRUE)", pruneDate)
+
+            let delete = NSBatchDeleteRequest(fetchRequest: pruneFetch)
+            delete.resultType = .resultTypeCount
+
+            do {
+                let result = try moc.execute(delete)
+                guard let deleteResult = result as? NSBatchDeleteResult else {
+                    DDLogError("EPC StorageManager: Could not read NSBatchDeleteResult")
+                    return
+                }
+
+                guard let count = deleteResult.result as? Int else {
+                    DDLogError("EPC StorageManager: Could not read NSBatchDeleteResult count")
+                    return
+                }
+                DDLogInfo("EPC StorageManager: Pruned \(count) events")
+
+            } catch let error {
+                DDLogError("EPC StorageManager: Error pruning events: \(error.localizedDescription)")
+            }
+        }
+    }
+
+    private func save(_ moc: NSManagedObjectContext) {
+        guard moc.hasChanges else {
+            return
+        }
+        do {
+            try moc.save()
+        } catch let error {
+            DDLogError("EPC: Error saving StorageManager managedObjectContext: \(error)")
+        }
+    }
+
+    private func performAndWait(_ block: (_ moc: NSManagedObjectContext) -> Void) {
+        let moc = self.managedObjectContext
+        moc.performAndWait {
+            block(moc)
+        }
+    }
+
+    private func perform(_ block: @escaping (_ moc: NSManagedObjectContext) -> Void) {
+        let moc = self.managedObjectContext
+        moc.perform {
+            block(moc)
+        }
+    }
+}
+
+extension StorageManager: PeriodicWorker {
+    public func doPeriodicWork(_ completion: @escaping () -> Void) {
```

Comment:
Yeah, I saw earlier that this wasn't firing and added it to the periodicWorkerController in WMFAppViewController.m; just hadn't pushed that update yet.  

The motivation here is that the main EPC class shouldn't be concerned with the storage manager's bookkeeping; all it should need to know how to do are (1) push events onto the queue, (2) pop events from the queue for submission, and (3) update enqueued events with `posted` or `failed` state.  But I suppose I can live with the main client directly calling StorageManager.pruneStaleEvents for now and cleaning up this (and some other implementation detail that's leaking in with the latest changes) at some possible future date.

Reply:
@mdholloway aah I see now. I'm also good with you doing it in `WMFAppViewController.m` instead in that case. Also to quote my earlier comment:

`Otherwise you could be marking a record as failed or posted after it has since been deleted, and I'm not sure that the underlying system would handle that well.`

This chaining might not be necessary once you work on the changes mentioned [here](https://github.com/wikimedia/wikipedia-ios/pull/3806#discussion_r547498524). Even if `markPosted` and `markFailed` runs on a record that has since been pruned, fetching the object via `existingObject` should let you ignore that error rather than crash.